drv/update-api/hiffy-rot-flash.sh

```diff
@@ -0,0 +1,72 @@
+#!/usr/bin/env bash
+
+# Load an image using the humility hiffy interface for the lpc55-update-server.
+#
+# Usage Example:
+# ARCHIVE=hubris/target/rot-carrier/dist/a/build-rot-carrier.zip rot-update-tools/hiffy-rot-flash.sh rot-carrier-v2/b/final.bin ImageB
+#
+# In the above example, `ARCHIVE` is set to the current image running on the
+# RoT. In this example, slot A is currently running, so we upload to slot b via
+# a saved build in `rot-carrier-v2` and additionally give the keyword `ImageB`.
+
+# NOTE: Make sure you are not runnning this on an ancient version of bash on Mac OSX
+
+set -x
+
+image=$1
+image_type=$2
+
+block_size=512
+binsize=`ls -la $image | cut -w -f 5`
+numblocks=$(("${binsize}" / "${block_size}"))
+total_before_last_block=$((numblocks * "${block_size}"))
+lastblocksize=$(("${binsize}" - "${total_before_last_block}"))
+loopend=$(("${numblocks}"-1))
+
+write_blocks() {
+  start="${1:?Missing start block number}"
+  shift
+  end="${1:?Missing end block number}"
+  shift
+  block_image="${1:?Missing path to image}"
+  shift
+
+  echo "${start}"
+  echo "${end}"
+
+  for (( i=start; i<=end; i++))
+  do
+    if ! humility -a "${ARCHIVE}" hiffy -c Update.write_one_block -a block_num="${i}" -i <(dd if="${block_image}" bs="${block_size}" count=1 skip="${i}")
+    then
+      # Retry once
+      #
+      # The following errors have been observed
+      #
+      # [2022-12-01T17:07:49Z WARN  probe_rs::config::target] Using custom sequence for LPC55S69
+      # humility: attached to 1fc9:0143:CW4N3IFK22UUX via CMSIS-DAP
+      # humility hiffy failed: A core architecture specific error occured
+      #
+      # Caused by:
+      #     0: Failed to read register DRW at address 0x0000000c
+      #     1: An error specific to the selected architecture occured
+      #     2: Target device responded with WAIT response to request.
+      # Error: humility failed
+      sleep 1
+      if !  humility -a "${ARCHIVE}" hiffy -c Update.write_one_block -a block_num="${i}" -i <(dd if="${block_image}" bs="${block_size}" count=1 skip="${i}")
+      then
+        exit 1
+      fi
+    fi
+  done
+}
+
+humility -a "${ARCHIVE}" hiffy -c Update.prep_image_update -a image_type="${image_type}"
+
+# Begin by invalidating the header which resides at offset 0x130 (in block zero).
+write_blocks 0 0 /dev/zero
+write_blocks 1 "${loopend}" "${image}"
+humility -a "${ARCHIVE}" hiffy -c Update.write_one_block -a block_num="${numblocks}" -i <(dd if="${image}" bs=1 count="${lastblocksize}" skip="${total_before_last_block}") 
```

Comment:
Because github is confusing, I assume you tagged line 68. The purpose is because the last block may not be of block size, and we only want to write. the correct number of bytes.

Reply:
In the case of the LPC55, only 512 blocks can be written. For STM32, you could conceivably have a partial block.
There should probably be a warning on writing of a partial block.
This command won't be necessary if there are an integral number of blocks.

In later work, we need to enforce the state of all the blocks in the bank.

Reply:
The lpc55-update-server pads smaller blocks with zeros, which prevents the need to edit the binary or pad the last block directly during the update. Now this may cause problems if an updater tried to send an intermediate block smaller than 512b, and would prevent the update from booting this is also detectable with a hash / signature check which we need to do anyway.